### DISCUSSION TITLE
How to replicate ImageJ rendering settings

### DISCUSSION THREAD

**Discussion by neale** [pyvista/pyvista#4067] [Q&A]
Hi everyone, I've been trying to use PyVista as a part of my workflow since other options seem to be either a closed ecosystem, or too slow. My usecase is to explore volumetric data at sizes >= (1024,1024,1024,3). If you've ever used imageJ you'll know that it has a hard time handling that scale, and its also hard to use as a part of any python based workflow. 

Which brings me to PyVista, I really liked it at first glance, the rendering is far smoother. I tinkered for a while, but I've noticed that the detail I get in the volume rendering is lower than I see in ImageJ, and it seems like its probably just a settings issue. Below is a comparison with something I saved out of ImageJ, with the same data rendered with PyVista. I'm wondering if anyone can help me dial in the rendering settings. 

![pyvista](https://user-images.githubusercontent.com/2781245/221497355-70c7824b-b0b5-4684-ad71-c6f7df0f8cf7.png)

The code used for loading and rendering the data:
```python
data = imread(path).compute()  # np.ndarray
rgb = data.reshape(-1, 3)  # [N, 3]
alpha = np.ones((*rgb.shape[:-1], 1), dtype=np.uint8)  # [N, 1]
scalars = np.concatenate((rgb, alpha), axis=-1) # [N, 4]

grid = pv.UniformGrid(dimensions=data.shape[:-1])

p = pv.Plotter(notebook=False, multi_samples=2)

pa = p.add_volume(
             grid,
             scalars=scalars, 
             mapper="smart", 
             blending='composite',
             shade=True)
p.show()
```

**Comment by banesullivan** [comment]
Could you share the data with us so that we can try to create a volume rendering that is consistent with the image you show?

**Reply by neale** [reply]
Of course, I put a `.tiff` file on google drive here: https://drive.google.com/drive/folders/1BZ1vtbyh3sfQHPk3Po53-BXzouaihLZt?usp=share_link
I think drive requires a sign-on, let me know if thats an issue.

**Comment by MatthewFlamm** [comment]
This looks to me like ImageJ is not doing a volume render in the sense of VTK and PyVista, but taking many isosurfaces and plotting them with transparency.  In ImageJ case, it doesn't even look that transparent, giving the layers a precision and solid look.  Whereas in PyVista, the volume render is projecting the volume data onto the pixels on the screen in a way that is different.

When trying to give an example to get something that looks like the ImageJ picture, I'm noticing something weird about the opacity keyword in `add_mesh`, which I will add a separate issue on.

**Reply by MatthewFlamm** [reply]
The weirdness was my forgetting to enable depth peeling.  This code gives me something that qualitatively looks like the ImageJ output to my eyes.  Note that this example requires `colorcet` for the more categorical colormap.

```py
import pyvista as pv
import numpy as np

data = pv.UniformGrid(dimensions=(20, 20, 20))
data["data"] = data.points[:, 0]**3 + data.points[:, 1] * data.points[:, 2]**2
min, max = (data["data"].min(), data["data"].max())
vals = np.linspace(min+0.001, max-0.001, 100)
surfs = pv.MultiBlock()
for val in vals:
    surfs.append(data.contour(isosurfaces=[val]))

plotter = pv.Plotter()
plotter.enable_depth_peeling()
plotter.add_mesh(surfs, scalars="data", opacity=0.5, cmap="glasbey_bw")
plotter.show()
```

![image](https://user-images.githubusercontent.com/39341281/221691479-de8139f1-2530-4a42-8338-2bbf7b82166f.png)

**Reply by neale** [reply]
@MatthewFlamm wow that looks closer! 
I'm trying to map this to what I've been doing in PyVista, and I admit I'm pretty confused about what's going on here. 
In principle I understand the guess about ImageJ plotting isosurfaces, and so here you're doing something similar with extracting contours and plotting a mesh. 
One thing that sticks out to me, is that the mesh is taking a `cmap`, where ideally I can use the RGB values I already have, which dictate the coloring.

**Reply by MatthewFlamm** [reply]
Do you read in the same tiff image to ImageJ and get that render?  Or is the tiff file an output of ImageJ? If the latter, it is very possible that ImageJ is doing something similar to what I did under the hood.  If the former, I'm not sure how ImageJ is contouring the data.  Contouring vector valued data is not very straightforward.  VTK only allows for contouring along one component of vector valued data (this is currently not exposed in PyVista).  Maybe someone else would have ideas here.

**Reply by neale** [reply]
The tiff volume is the output of some other process. ImageJ is rendering it, and then I can click the "screenshot" button :/

**Reply by neale** [reply]
I'm not after an ImageJ clone per-se, but the settings I was using to get a render with PyVista was just not looking good. I'd take anything that was comparable in terms of visual quality. Placement in the python ecosystem is worth a lot.
